File: 19_graph/q2_3.py
import sys
sys.setrecursionlimit(10000)


def dfs(graph, node_id, visited, depth, depth_list):
    if visited[node_id]:
        depth_list[node_id] = depth
        return

    visited[node_id] = True

    adj_id_list = graph[node_id]
    for adj_id in adj_id_list:
        if not visited[adj_id]:
            depth_list[adj_id] = depth + 1
            dfs(graph, adj_id, visited, depth + 1, depth_list)


n = int(input())
p_list = [int(v) for v in input().split()]
graph = [list() for _ in range(n)]

for i in range(1, n):
    parent_node_id = p_list[i - 1]
    # 親から子へのエッジを貼れば良いらしい。
    # graph[i].append(parent_node_id)
    graph[parent_node_id].append(i)

visited = [False] * n
depth_list = [0] * n

depth_list[0] = 0
dfs(graph, 0, visited, depth_list[0], depth_list)

for i in range(n):
    print(depth_list[i])

File: 19_graph/q2_3.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <string>
#include <sstream>
#include <algorithm>

typedef unsigned long long ull;
typedef unsigned long ul;

#define rep(T, i, n) for (T i = 0; i < (T)(n); i++)
#define Graph(T) std::vector<std::vector<T>>

template <class T>
void GetSplitStringVec(const std::string& raw_data, std::vector<T>* v) {
  // ref split string: https://marycore.jp/prog/cpp/std-string-split/
  std::stringstream ss{raw_data};
  std::string buf;
  while (std::getline(ss, buf, ' ')) {
    T data;
    std::istringstream iss(buf);
    iss >> data;
    v->push_back(data);
  }
}

template <class T>
const T dfs(const Graph(T) & graph, const T& node_id, const T& depth,
            std::vector<bool>* visited) {
  T ret_depth = depth;
  if (visited->at(node_id)) return ret_depth;
  visited->at(node_id) = true;

  for (auto adj_node_id : graph[node_id]) {
    if (visited->at(adj_node_id)) continue;
    // ret_depth = std::max(ret_depth, dfs<T>(graph, adj_node_id, depth + 1, visited));
    ret_depth = dfs<T>(graph, adj_node_id, depth + 1, visited);
  }
  return ret_depth;
}

int main() {
  std::string raw_str;
  std::getline(std::cin, raw_str);
  ul n = std::stoul(raw_str);

  std::vector<ul> p_list;
  std::getline(std::cin, raw_str);
  GetSplitStringVec<ul>(raw_str, &p_list);

  Graph(ul) graph(n);
  rep(ul, i, n - 1) {
    // 子から親への参照を登録
    graph[i + 1].push_back(p_list[i]);
  }

  // 子から根までの深さを調べる。
  std::cout << 0 << std::endl;  // 根は深さ０
  rep(ul, i, n - 1) {
    std::vector<bool> visited(n, false);
    ul depth = dfs<ul>(graph, i + 1, 0, &visited);
    std::cout << depth << std::endl;
  }

  return 0;
}
